File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, h = 0, len(nums) - 1

        while l < h:
            m = (l + h) // 2

            if nums[m] > nums[h]:
                l = m + 1
            else:
                h = m
        
        return nums[l]

File: Data Structures & Algorithms/maximum-product-subarray/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxProduct(nums) {
        let globalMax = nums[0];
        let currMax = nums[0];
        let currMin = nums[0];

        for (let i = 1; i < nums.length; i++) {
            const num = nums[i];

            const tempMax = Math.max(num, currMax * num, currMin * num);
            currMin = Math.min(num, currMax * num, currMin * num);

            currMax = tempMax;
            globalMax = Math.max(globalMax, currMax);
        }

        return globalMax;
    }
}
